azurelinuxagent/common/cgroup.py

```diff
@@ -118,76 +118,72 @@ def is_active(self):
 
 class CpuCgroup(CGroup):
     def __init__(self, name, cgroup_path):
-        """
-        Initialize _data collection for the Cpu controller. User must call update() before attempting to get
-        any useful metrics.
-
-        :return: CpuCgroup
-        """
         super(CpuCgroup, self).__init__(name, cgroup_path, "cpu")
 
         self._osutil = get_osutil()
-        self._current_cpu_total = 0
-        self._previous_cpu_total = 0
-        self._current_system_cpu = self._osutil.get_total_cpu_ticks_since_boot()
-        self._previous_system_cpu = 0
+        self._previous_cgroup_cpu = None
+        self._previous_system_cpu = None
+        self._current_cgroup_cpu = None
+        self._current_system_cpu = None
 
     def __str__(self):
         return "cgroup: Name: {0}, cgroup_path: {1}; Controller: {2}".format(
             self.name, self.path, self.controller
         )
 
-    def _get_current_cpu_total(self):
+    def _get_cpu_ticks(self):
         """
-        Compute the number of USER_HZ of CPU time (user and system) consumed by this cgroup since boot.
-
-        :return: int
+        Returns the number of USER_HZ of CPU time (user and system) consumed by this cgroup.
         """
-        cpu_total = 0
         try:
             cpu_stat = self._get_file_contents('cpuacct.stat')
         except Exception as e:
             if isinstance(e, (IOError, OSError)) and e.errno == errno.ENOENT:
                 raise e
-            raise CGroupsException("Exception while attempting to read {0}".format("cpuacct.stat"), e)
+            raise CGroupsException("Failed to read cpuacct.stat: {0}".format(ustr(e)))
+
+        cpu_ticks = 0
 
         if cpu_stat:
             m = re_user_system_times.match(cpu_stat)
             if m:
-                cpu_total = int(m.groups()[0]) + int(m.groups()[1])
-        return cpu_total
+                cpu_ticks = int(m.groups()[0]) + int(m.groups()[1])
+
+        return cpu_ticks
 
-    def _update_cpu_data(self):
+    def _cpu_usage_initialized(self):
+        return self._current_cgroup_cpu is not None and self._current_system_cpu is not None
+
+    def initialize_cpu_usage(self):
```

Comment:
Can we call `initialize_cpu_usage` within the `__init__` itself?

Reply:
The get*cgroup functions instantiate cgroups that may never be tracked... I thought it was not worth initializing the cpu usage for those (especially since the actual cgroups may not exist on disk yet --- if we initialize on init the we would need to ensure the cgroups exist before calling the getter).